ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java

```diff
@@ -5167,7 +5167,14 @@ private RelNode genGBHavingLogicalPlan(QB qb, RelNode srcRel) throws SemanticExc
               UnsupportedFeature.Having_clause_without_any_groupby);
         }
         ASTNode targetNode = (ASTNode) havingClause.getChild(0);
-        validateNoHavingReferenceToAlias(qb, targetNode);
+        QBParseInfo qbPI = qb.getParseInfo();
+        Map<ASTNode, String> exprToAlias = qbPI.getAllExprToColumnAlias();
+        RowResolver inputRR = relToHiveRR.get(srcRel);
+        for (ASTNode astNode : exprToAlias.keySet()) {
+          if (inputRR.getExpression(astNode) != null) {
+            inputRR.put("", exprToAlias.get(astNode), inputRR.getExpression(astNode));
+          }
+        }
```

Comment:
This part is basically modifying the `RowResolver` adding some kind of reverse mappings. Moreover, this logic is duplicated in `SemanticAnalyzer#genHavingPlan`. I think it makes sense to refactor this snippet and create a new method in the `RowResolver` class with a proper name (and javadoc if necessary) that clarifies what is the code supposed to do. The only parameter to the method should be a `Map<ASTNode, String>`.